lightning/src/ln/channelmanager.rs

```diff
@@ -4284,6 +4286,61 @@ where
 			)
 	}
 
+	#[cfg(async_payments)]
+	fn initiate_async_payment(
+		&self, invoice: &StaticInvoice, payment_id: PaymentId
+	) -> Result<(), Bolt12PaymentError> {
+		let mut res = Ok(());
+		PersistenceNotifierGuard::optionally_notify(self, || {
+			let outbound_pmts_res = self.pending_outbound_payments.static_invoice_received(
+				invoice, payment_id, &*self.entropy_source, &self.pending_events
+			);
+			let payment_release_secret = match outbound_pmts_res {
+				Ok(secret) => secret,
+				Err(Bolt12PaymentError::UnexpectedInvoice) | Err(Bolt12PaymentError::DuplicateInvoice) => {
+					res = outbound_pmts_res.map(|_| ());
+					return NotifyOption::SkipPersistNoEvents
+				},
+				Err(e) => {
+					res = Err(e);
+					return NotifyOption::DoPersist
+				}
+			};
+
+			let reply_paths = match self.create_blinded_paths(
+				MessageContext::AsyncPayments(AsyncPaymentsContext::OutboundPayment { payment_id })
+			) {
+				Ok(paths) => paths,
+				Err(()) => {
+					self.abandon_payment_with_reason(payment_id, PaymentFailureReason::RouteNotFound);
+					res = Err(Bolt12PaymentError::SendingFailed(RetryableSendFailure::RouteNotFound));
+					return NotifyOption::DoPersist
```

Comment:
Should we create a separate error for this? This is used for failing to find a route to the recipient for a payment, but here it is used when failing to find a route to the sender for a message.

Reply:
How does updating the docs sound? Unless you think the separate error could result in a separate course of action for the user?

Reply:
Hmm... arguably we should never return a `Bolt12PaymentError::SendingFailed` variant here since a send is never initiated. But that would mean using something other than `Bolt12PaymentError`. It feels like we're abusing the error type here -- though possibly elsewhere already -- but I'm not sure the alternative of creating similar but distinct error types is any better.

@TheBlueMatt Not sure if you have any strong opinions here. In https://github.com/lightningdevkit/rust-lightning/pull/3175#discussion_r1706024322, we decided on using separate error types, though there was a logical inbound/outbound separation there. Here at least everything is outbound, so maybe using a single error type is fine?

Reply:
Note that #3175 does fix the `pay_for_offer` case. I'd lean towards adding another variant to `Bolt12PaymentError`, even though it wouldn't be applicable for non-static invoices. The `Bolt12PaymentError::SendingFailed` variant is also still not relevant here, either, for the reason I gave earlier. Though fixing that would mean creating a new error type that is essentially `Bolt12PaymentError` minus `SendingFailed` plus the new variant. 🙁

Reply:
Updated `Bolt12PaymentError` to have a new variant for this case. 

We also currently abandon the payment with `PaymentFailureReason::RouteNotFound`, which also doesn't match this case per its docs. So most likely `PaymentFailureReason` could use a new variant for this case too. Pre-existing to this PR, `::RouteNotFound` also could use a docs update for its use in `send_payment_for_bolt12_invoice`. Let me know if that all makes sense, though I may make these changes in the follow-up #3207 since this PR is growing.

Reply:
> Updated `Bolt12PaymentError` to have a new variant for this case.
> 
> We also currently abandon the payment with `PaymentFailureReason::RouteNotFound`, which also doesn't match this case per its docs. So most likely `PaymentFailureReason` could use a new variant for this case too. Pre-existing to this PR, `::RouteNotFound` also could use a docs update for its use in `send_payment_for_bolt12_invoice`. Let me know if that all makes sense, though I may make these changes in the follow-up #3207 since this PR is growing.

Yeah, let's add a new `PaymentFailureReason` variant and add a `legacy_reason` conversion, I guess? Regarding a doc update, is that largely because it says, "while retrying the payment"?

Reply:
Yeah, the phrasing makes the variant sound exclusive to retries at the moment.